src/it/scala/io/iohk/ethereum/sync/FastSyncItSpec.scala

```diff
@@ -0,0 +1,463 @@
+        _ <- peer3.importNBlocksToTheTopForm(peer3.getCurrentState(), 1000)(IdentityUpdate)
+        _ <- peer1.connectToPeers(Set(peer2.node, peer3.node))
+        _ <- peer1.startFastSync()
+        _ <- peer1.waitForFastSyncFinish()
+      } yield {
+        assert(peer1.bl.getBestBlockNumber() == peer2.bl.getBestBlockNumber() - peer2.testSyncConfig.targetBlockOffset)
+        assert(peer1.bl.getBestBlockNumber() == peer3.bl.getBestBlockNumber() - peer3.testSyncConfig.targetBlockOffset)
+      }
+  }
+
+  it should "should sync blockchain with state nodes" in customTestCaseResourceM(FakePeer.start3FakePeersRes()) {
+    case (peer1, peer2, peer3) =>
+      for {
+        _ <- peer2.importNBlocksToTheTopForm(peer2.getCurrentState(), 1000)(updateStateAtBlock(500))
+        _ <- peer3.importNBlocksToTheTopForm(peer3.getCurrentState(), 1000)(updateStateAtBlock(500))
+        _ <- peer1.connectToPeers(Set(peer2.node, peer3.node))
+        _ <- peer1.startFastSync()
+        _ <- peer1.waitForFastSyncFinish()
+      } yield {
+        val trie = peer1.getBestBlockTrie()
+        // due to the fact that function generating state is deterministic both peer2 and peer3 ends up with exactly same
+        // state, so peer1 can get whole trie from both of them.
+        assert(peer1.bl.getBestBlockNumber() == peer2.bl.getBestBlockNumber() - peer2.testSyncConfig.targetBlockOffset)
+        assert(peer1.bl.getBestBlockNumber() == peer3.bl.getBestBlockNumber() - peer3.testSyncConfig.targetBlockOffset)
+        assert(trie.isDefined)
+      }
+  }
+
+
+  it should "should update target block" in customTestCaseResourceM(FakePeer.start2FakePeersRes()) {
+    case (peer1, peer2) =>
+      for {
+        _ <- peer2.importNBlocksToTheTopForm(peer2.getCurrentState(), 1000)(IdentityUpdate)
+        _ <- peer1.connectToPeers(Set(peer2.node))
+        _ <- peer2.importBlocksUntil(2000)(IdentityUpdate).startAndForget
+        _ <- peer1.startFastSync()
+        _ <- peer1.waitForFastSyncFinish()
+      } yield {
+        assert(peer1.bl.getBestBlockNumber() == peer2.bl.getBestBlockNumber() - peer2.testSyncConfig.targetBlockOffset)
+      }
+  }
+}
+
+object FastSyncItSpec {
+  private def retryUntilWithDelay[A](source: Task[A], delay: FiniteDuration, maxRetries: Int)(
+    predicate: A => Boolean
+  ): Task[A] = {
+    source.delayExecution(delay).flatMap { result =>
+      if (predicate(result)) {
+        Task.now(result)
+      } else {
+        if (maxRetries > 0) {
+          retryUntilWithDelay(source, delay, maxRetries - 1)(predicate)
+        } else {
+          Task.raiseError(new TimeoutException("Task time out after all retries"))
+        }
+      }
+    }
+  }
+
+  def randomAddress(): InetSocketAddress = {
+    val s = new ServerSocket(0)
+    try {
+      new InetSocketAddress("localhost", s.getLocalPort)
+    } finally {
+      s.close()
+    }
+  }
+
+  def customTestCaseResourceM[T](fixture: Resource[Task, T])
```

Comment:
Minor: should we move this somewhere for re-using it?

Reply:
I saw someone porting whole `SpecBase` from other project, i will make use of it